Cloud Validator Documentation (#782)
ParsePlatform/Docs: _includes/cloudcode/cloud-code.md

    sum += results[i].get("stars");
  }
  return sum / results.length;
},{

>   fields:['movie']

Reviewer: I'm guessing this was a mistake...
```suggestion
```

Author: I figured it might be an easy way to warm into the cloud validation. Considering `request.params.movie` is used in the query, I thought it would make sense to make sure it's not `null`.

Reviewer: I think it's confusing because it isn't explained so if someone doesn't look at the cloud validation section they may not have any idea what it does.

}
```

## Implementing cloud function validation

*Available only on parse-server cloud code starting 4.-.-*

> It's important to make sure the parameters provided to a Cloud function are provided, and are in the necessary format. Starting with Parse Server 4.-.-, you can now specify a validator function or object which will be called prior to your cloud function.

Reviewer: This doesn't read too well for me... maybe it's better like this...
```suggestion
It's important to make sure the parameters required for a Cloud function are provided, and are in the necessary format. Starting with Parse Server 4.-.-, you can now specify a validator function or object which will be called prior to your cloud function.
```

Let's take a look at the `averageStars` example. If you wanted to make sure that `request.params.movie` is provided, and `averageStars` can only be called by logged in users, you could add a validator object to the function.

```javascript
Parse.Cloud.define("averageStars", async (request) => {
  const query = new Parse.Query("Review");
  query.equalTo("movie", request.params.movie);
  const results = await query.find();
  let sum = 0;
  for (let i = 0; i < results.length; ++i) {
    sum += results[i].get("stars");
  }
  return sum / results.length;
},{
  fields : ['movie'],
  requireUser: true
});
```

> If the rules specified in the validator object aren't met, the cloud function won't run. This means that you can confidently build your trigger, knowing that `request.params.movie` is defined, as well as `request.user`.

Reviewer: ```suggestion
If the rules specified in the validator object aren't met, the Cloud Function won't run. This means that you can confidently build your function, knowing that `request.params.movie` is defined, as well as `request.user`.
```

> ## More Advanced Validation

Reviewer: Should be under the 'Implementing cloud function validation' section I think...
```suggestion
### More Advanced Validation
```

> *Available only on parse-server cloud code starting 4.-.-*

Reviewer: As it's mentioned above, if it's in the same section this should be unnecessary...
```suggestion
```

> Often, not only is it important that `request.params.movie` is defined, but also that it is the correct data type. You can do this by providing an `Object` to the `fields` parameter in the Validator.

Reviewer: ```suggestion
Often, not only is it important that `request.params.movie` is defined, but also that it's the correct data type. You can do this by providing an `Object` to the `fields` parameter in the Validator.
```

- `request.user` is defined
- `request.user.get('accType')` is defined
- `request.user.get('accType')` is equal to 'reviewer'

However, the requested user could set 'accType' to reviewer, and then recall the function. Here, you could provide validation on a `Parse.User` `beforeSave` trigger. `beforeSave` validators have a few additional options available, to help you make sure your data is secure.

```javascript
Parse.Cloud.beforeSave(Parse.User, () => {
  // any additional beforeSave logic here
}, {
    fields: {
      accType: {
        default: 'viewer',
        constant: true
      },
    },
});
```
This means that the field `accType` on `Parse.User` will be 'viewer' on signup, and will be unchangable, unless `masterKey` is provided.

The full range of Built-In Validation Options are:

- `requireMaster`, whether the function requires a `masterKey` to run
- `requireUser`, whether the function requires a `request.user` to run
- `validateMasterKey`, whether the validator should run on `masterKey` (defaults to false)
- `fields`, an `Array` or `Object` of fields that are required on the request.
- `requireUserKeys`, an `Array` of fields to be validated on `request.user`

> The full range of Built-In Validation Options on `.fields` are:

Reviewer: ```suggestion
The full range of built-in Validation Options are:

- `requireMaster`: whether the function requires a `masterKey` to run.
- `requireUser`: whether the function requires a `request.user` to run.
- `validateMasterKey`: whether the validator should run on `masterKey` (defaults to false).
- `fields`: an `Array` or `Object` of fields that are required on the request.
- `requireUserKeys`: an `Array` of fields to be validated on `request.user`.

The full range of built-in Validation Options on `.fields` are:
```

- `type`, the type of the `request.params[field]` or `request.object.get(field)`
- `default`, what the field should default to if it's `null`,
- `required`, whether the field is required.
- `options`, a singular option, array of options, or custom function of allowed values for the field.
- `constant`, whether the field is immutable.

> - `error`, a custom error message if validation fails.

Reviewer: ```suggestion
- `type`: the type of the `request.params[field]` or `request.object.get(field)`.
- `default`: what the field should default to if it's `null`.
- `required`: whether the field is required.
- `options`: a singular option, array of options, or custom function of allowed values for the field.
- `constant`: whether the field is immutable.
- `error`: a custom error message if validation fails.
```

> You can also pass a function to the Validator. This can help you apply reoccuring logic to your cloud code.

Reviewer: ```suggestion
You can also pass a function to the Validator. This can help you apply reoccuring logic to your Cloud Code.
```

```javascript
const validationRules = request => {
  if (request.master) {
    return;
  }
  if (!request.user || request.user.id !== 'masterUser') {
    throw 'Unauthorized';
  }
}

Parse.Cloud.define('adminFunction', request => {
// do admin code here, confident that request.user.id is masterUser, or masterKey is provided
},validationRules)

> Parse.Cloud.define('adminFunctionTwo', request => {
> // do admin code here, confident that request.user.id is masterUser, or masterKey is provided
> },validationRules)
> 

Reviewer: I imagine this is a mistake...
```suggestion
```

Author: I wanted to reiterate that you can reuse `validationRules` and pass to multiple functions, but can remove!

Reviewer: Tbh I thought you had just pasted it twice by mistake (I didn't notice the functions had different names). I see where you are coming from, maybe it can stay as is.

```

### Some considerations to be aware of

> - The validation function will run prior to your cloud code functions. You can use async and promises here, but try to keep the validation as simple and fast as possible so your cloud requests resolve in the shortest time.

Reviewer: ```suggestion
- The validation function will run prior to your Cloud Code Functions. You can use async and promises here, but try to keep the validation as simple and fast as possible so your cloud requests resolve quickly.
```

- As previously mentioned, cloud validator objects will not validate if a masterKey is provided, unless `validateMasterKey:true` is set. However, if you set your validator to a function, the function will **always** run.

> This range of options should help you write more secure cloud code. If you need help in any way, feel free to reach out at our [developer supported community forum](https://community.parseplatform.org/).

Reviewer: ```suggestion
This range of options should help you write more secure Cloud Code. If you need help in any way, feel free to reach out on our [developer supported community forum](https://community.parseplatform.org/).
```

## beforeSave

### Implementing data validation

Another reason to run code in the cloud is to enforce a particular data format. For example, you might have both an Android and an iOS app, and you want to validate data for each of those. Rather than writing code once for each client environment, you can write it just once with Cloud Code.

Let's take a look at our movie review example. When you're choosing how many stars to give something, you can typically only give 1, 2, 3, 4, or 5 stars. You can't give -6 stars or 1337 stars in a review. If we want to reject reviews that are out of bounds, we can do this with the `beforeSave` method:

```javascript
Parse.Cloud.beforeSave("Review", (request) => {
// do any additional beforeSave logic here
},{
  fields: {
    stars : {
      required:true,
      options: stars => {
        return stars >= 1 && stars =< 5;
      },
      error: 'Your review must be between one and five stars'

>     }

Reviewer: I'm not sure whether it's a good idea to completely remove examples of the 'old way' of validation before the feature is battle tested.

Perhaps here there should be another example with how the code would look with the new validation feature or a link to the section you have added above.

@davimacedo @mtrezza Any thoughts on this?

Reviewer 2: maybe we could should the two ways to do that

Author: How about we leave the current method, and add the new logic below?

E.g:

```javascript
Parse.Cloud.beforeSave("Review", (request) => {
  if (request.object.get("stars") < 1) {
    throw "you cannot give less than one star";
  }

  if (request.object.get("stars") > 5) {
    throw "you cannot give more than five stars";
  }
});
```

Or, using a Cloud Validator:

```javascript
Parse.Cloud.beforeSave("Review", (request) => {
// do any additional beforeSave logic here
},{
  fields: {
    stars : {
      required:true,
      options: stars => {
        return stars >= 1 && stars =< 5;
      },
      error: 'Your review must be between one and five stars'
    }
  }
});

```

Reviewer 2: Yes. It works for me.

Reviewer: I'm happy with that.

```javascript
Parse.Cloud.beforeSave(Parse.User, async (request) => {
    // code here
},
  // Validation Object or Validation Function

> )

Reviewer: I think it could be confusing having this here.
